sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -769,6 +770,27 @@ object CombineFilters extends Rule[LogicalPlan] {
 }
 
 /**
+ * Remove all the deterministic conditions in a [[Filter]] that are contained in the Child.
+ */
+object PruneFilters extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Filter(fc, p: LogicalPlan) =>
+      val (prunedPredicates, remainingPredicates) =
+        splitConjunctivePredicates(fc).partition { cond =>
+          cond.deterministic && p.constraints.contains(cond)
+        }
+      if (prunedPredicates.isEmpty) {
+        f
+      } else if (remainingPredicates.isEmpty) {
+        p
+      } else {
+        val newCond = remainingPredicates.reduceOption(And).getOrElse(Literal(true))
+        Filter(newCond, p: LogicalPlan)
```

Comment:
I don't think that you need the type ascription (`: LogicalPlan`) here or above in the match.